source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1877,24 +1884,40 @@ void COpenGLDriver::SAuxContext::flushState(GL_STATE_BITS stateBits)
     core::smart_refctd_ptr<COpenGLRenderpassIndependentPipeline> prevPipeline = currentState.pipeline;
     if (stateBits & GSB_PIPELINE_AND_RASTER_PARAMETERS)
     {
-        if (nextState.pipeline != currentState.pipeline) {
-            SGraphicsPipelineHash hash;
-            for (uint32_t i = 0u; i < COpenGLRenderpassIndependentPipeline::SHADER_STAGE_COUNT; ++i)
+        if (nextState.pipeline != currentState.pipeline)
+        {
+            if (nextState.usedShadersHash != currentState.usedShadersHash)
             {
-                hash[i] = currentState.pipeline->getShaderAtIndex(i) ?
-                    static_cast<const COpenGLSpecializedShader*>(currentState.pipeline->getShaderAtIndex(i))->getGLnameForCtx(this->ID) :
-                    0u;
-            }
+                GLuint GLname = 0u;
 
-            GLuint GLname = 0u;
-            auto found = GraphicsPipelineMap.find(hash);
-            if (found != GraphicsPipelineMap.end())
-                GLname = found->second;
-            else
-                GLname = createGraphicsPipelineInCache(hash);
+                constexpr SOpenGLState::SGraphicsPipelineHash NULL_HASH = {0u, 0u, 0u, 0u, 0u};
 
-            extGlUseProgram(0);//TODO hm??
-            extGlBindProgramPipeline(GLname);
+                HashPipelinePair lookingFor{nextState.usedShadersHash, {}};
+                if (lookingFor.first != NULL_HASH)
+                {
+                    auto found = std::lower_bound(GraphicsPipelineMap.begin(), GraphicsPipelineMap.end(), lookingFor);
+                    if (found != GraphicsPipelineMap.end() && found->first==nextState.usedShadersHash)
+                    {
+                        GLname = found->second.GLname;
+                        found->second.lastValidated = CNullDriver::ReallocationCounter;
```

Comment:
I don't think anyone needs the `lastValidated` field anymore, since gfx pipeline is immutable now, so vertex formats, shaders, etc. can't change.

Unless you're trying to avoid vertex buffer binds (hard optimization)

Reply:
actually it's "last bound" time-stamp or whatever ReallocationCounter gives, need for lru-like thing

Reply:
ah ok.